Rewrite using JSON3(read, Type)

evetion opened the pull request:
Will be a severe breaking release as almost everything changed, but should also offer great performance.

- [x] Fix Feature(Collection) api/tests
- [x] ~~Needs convert/construction logic from Tables/GeoInterface~~ other PRs.
- [x] Figure id, crs, etc.
- [x] Precompile most calls by using SnoopCompile on test samples

Our previous code was much more flexible in allowing anything to be parsed as GeoJSON, but we will miss out on that, but gain performance. Setting Feature(Collection) to mutable does allow some flexibility, such as not writing the bbox or other attributes if they're nothing. It does double the compilation time, but is almost on par once compiled. I can imagine we add other nullable attributes such as id and crs to be slightly backwards compatible.

Added a dep on ComputedFieldTypes to calculate the type of the bounding box, but it might be overkill.

evetion commented:
- [ ] Do we want to subtype AbstractArray for Geometry?

rafaqz commented:
Would have been nice to keep the same files so we can see the changes ;)

rafaqz commented on src/geointerface.jl:
```diff
@@ -1,49 +1,47 @@
 GI.ngeom(::GI.LineStringTrait, g::LineString) = length(g)
-GI.getgeom(::GI.LineStringTrait, g::LineString, i::Integer) = Point(coordinates = g[i])
-GI.getpoint(::GI.LineStringTrait, g::LineString, i::Int) = Point(coordinates = g[i])
+GI.getgeom(::GI.LineStringTrait, g::LineString, i::Integer) = Point("Point", g[i])
```
Why do we need the "Point" string? Don't we know this from the type?

rafaqz commented on src/geojson_types.jl:
```diff
@@ -0,0 +1,184 @@
+abstract type AbstractFeatureCollection{D} <: GeoJSONT{D} end
+
+struct Point{D} <: AbstractGeometry{D}
+    type::String
```
This could be particularly costly in memory because we are allocating a string for every single point. Is there a way to throw that information away, and get it back from a function of the type?

We could also use https://github.com/JuliaString/InternedStrings.jl to avoid the allocations

evetion commented:
> Would have been nice to keep the same files so we can see the changes ;)

I tried, but failed. Essentially the following happened:
- json.jl => io.jl
- geometries.jl, features.jl => geojson_types.jl

rafaqz commented on src/geojson_types.jl:
```diff
@@ -0,0 +1,184 @@
+    return val, state
+end
+
+@computed mutable struct Feature{D} <: GeoJSONT{D}
```
Maybe make a note on what `@computed` does here

rafaqz commented on src/geojson_types.jl:
```diff
@@ -0,0 +1,184 @@
+properties(f::Feature) = f.properties
+coordinates(f::Feature) = coordinates(geometry(f))
+Base.show(io::IO, x::Feature{D}) where {D} = print(io, "Feature with $(D)D $(x.geometry.type) geometry and $(length(x.properties)) properties")
+
```
Docstrings for structs would be nice

rafaqz commented on src/geojson_types.jl:
```diff
@@ -0,0 +1,184 @@
+
+Base.show(io::IO, x::LazyFeatureCollection) = print(io, "LazyFeatureCollection with $(length(x.features)) features")
+Base.getindex(x::LazyFeatureCollection{D}, i::Int) where {D} = JSON3.read(codeunits(x.features[i]), Feature{D})::Feature{D}
+
```
Docstring

rafaqz commented on src/geojson_types.jl:
```diff
@@ -0,0 +1,184 @@
+Base.IndexStyle(::AbstractFeatureCollection) = IndexLinear()
+
+bbox(fc::AbstractFeatureCollection) = fc.bbox
+
```
Docstring and comment on `@computed`

rafaqz commented on src/geojson_types.jl:
```diff
@@ -0,0 +1,184 @@
+StructTypes.construct(::Type{LazyFeature{D}}, x::JSON3.RawValue) where {D} = LazyFeature{D}(x.bytes, x.pos, x.len)
+Base.codeunits(x::LazyFeature) = unsafe_string(pointer(x.bytes, x.pos), x.len)
+JSON3.rawbytes(x::LazyFeature) = codeunits(x)
+
```
Docstring and comment on `@computed`

rafaqz commented on src/geojson_types.jl:
```diff
@@ -0,0 +1,184 @@
+abstract type GeoJSONT{D} end
+abstract type AbstractGeometry{D} <: GeoJSONT{D} end
+abstract type AbstractFeatureCollection{D} <: GeoJSONT{D} end
+
```
Docstring here and for other types

rafaqz commented on src/geojson_types.jl:
```diff
@@ -162,6 +219,27 @@ StructTypes.StructType(::Type{<:GeoJSONWrapper}) = StructTypes.CustomStruct()
 StructTypes.lowertype(::Type{<:GeoJSONWrapper{D}}) where {D} = GeoJSONT{D}
 
+const point = "Point"
+type(::Type{<:Point}) = point
```
Can we call this `stringtype` ?

evetion commented:
Ok, so this requires the unreleased GeoInterface for getting the extent (using `trait` instead of `geomtrait`, so it also works on FeatureCollections).

The remaining error is from Aqua, which complains about Unbound type parameters, in all the constructors for Geometry, like:
```julia
GeoJSON.Polygon(bbox::Union{Nothing, Vector{Float32}}, coordinates::Union{Nothing, Array{Array{Tuple{Vararg{Float32, D}}, 1}, 1}}) where D @ GeoJSON ~/code/GeoJSON.jl/src/geojson_types.jl:46
```

I'm not sure what the unbound problem is, so help would be appreciated.

The remaining thing is to improve the shadowing of fields vs keys in the properties of Feature(Collections), as we also need to add `id` and `crs` to those exclusions.

I also still need to write a changelog with all the breaking changes in this release.
As of now I have:

- The keys of the properties of a Feature(Collections) are now `Symbol`s.
- All Type signatures changed, everything is now typed, so no more JSON3.Array and JSON3.Object
- Custom attributes (apart from `id`, `bbox` and `crs`) are now unsupported and ignored.
- Geometries are not AbstractVectors anymore

rafaqz commented:
The unbound problem is with the `Union` - there is no `D` if the type is `Nothing`. Maybe we have to remove the possibility to have nothing and use an empty vector instead.

evetion commented:
> I think there is no way around providing D manually.

Well you can do

```julia
Point(; bbox=nothing, coordinates::NTuple{D,T}) where {D,T<:Real} = Point{D}(bbox, coordinates)
```

But then you have unbound T. Otherwise you have to hardcode the Type to Float32, or change the signature from {D}, to {D,T} for everything. What do you think?

rafaqz commented:
Don't we know `D` anyway? Why not use it? Maybe I need to read over the code

evetion commented:
This is mainly for end users, to easily construct a thing with kw arguments. And D can be inferred when `coordinates` is passed.

evetion commented:
Ok, I've now set {D,T} as the type parameters. Sadly, I still get the unbound errors. Another thing, precompile doesn't seem to do much at all, even with SnoopPrecompile. Reading in the 24MB always takes 2.5s at the first time, afterwards it drops to 0.04s. Maybe we need to place a bit of JSON3 reads just in the module itself to warm things up a bit. Or make a PR over at JSON3 to precompile much more types?

Anyway, I'd appreciate a deeper review, as this is close to complete.

rafaqz commented:
Ok well I didn't recommend adding `T` ;)

`D` and `nothing` is the problem... If there is no `D` we'll have to make it up (e.g. 2) and it has to be in a separate method to the one where D is known to avoid the unbound types

But personally I would force everyone to use D, or like GeometryBasics.jl, add `Polygon2` as a shortcut to that. They have the same problem.

I'll try review tomorrow or wednesday

evetion commented:
After some experiments with precompile/snoopprecompile, I just added a lot of precompile statements from --trace-compile. This is the only reliable method that got initial times down for parsing the 24MB featurecollection from GeoMakie.

Also fixed the unbound parameters test by defaulting to 2d for `Point(coordinates=(1,2))` construction. Nice thing about adding T is that we essentially added back numbertype support back to GeoJSON and people can choose between Int, Float16 etc.

rafaqz commented on src/geojson_types.jl:
```diff
@@ -0,0 +1,350 @@
+Should not be used, as it is not part of the GeoJSON specification.
+The CRS of a GeoJSON object is always WGS84.
+"""
+struct CRS
```
Hmm... why does this need to exist?

evetion replied:
I believe there's a test example that has a crs. I tried to be somewhat backwards compatible in terms of the extra fields encountered in the old tests.

rafaqz replied:
Yeah it probably makes sense. Although we will need to explicitly not handle crs point order.

rafaqz commented on src/precompile.jl:
```diff
@@ -0,0 +1,196 @@
+precompile(Tuple{Type{Union{}}})
```
Is `precompile` still the way to do this? a lot of packages are just running the function. At least you know it it breaks that way.

visr replied:
SnoopPrecompile's `@precompile_all_calls` might be a good idea https://timholy.github.io/SnoopCompile.jl/dev/snoop_pc/

evetion replied:
Well, running is also fine, but you can only do that on smaller packages geojsons, and the test geojson we package didn't actually decrease the runtime much if at all for our 24MB usecase. So I think this is the best. It's also easy to make changes to this code later on or even remove it.

The tests will be responsible to detect breakage.

evetion replied:
I actually ran with @precompile_all_calls first, but that didn't seem to do much either. Might be my setup.

rafaqz commented on src/geojson_types.jl:
```diff
@@ -0,0 +1,350 @@
+@inline StructTypes.lowertype(::Type{<:GeoJSONWrapper{D,T}}) where {D,T} = GeoJSONT{D,T}
+
+const point = "Point"
+@inline typestring(::Type{<:Point}) = point
```
I doubt these `@inline` macros are required, this will compile away. 

And is there a reason to not just return the string from the function?

rafaqz commented on src/geojson_types.jl:
```diff
@@ -0,0 +1,350 @@
+GeoJSONWrapper{D,T}(obj::X) where {D,T,X<:GeoJSONT{D,T}} = GeoJSONWrapper{D,T,X}(obj)
+
+# symbol (from json string type) to struct mapping
+function geom_mapping(D, T)
```
May need to make sure this specialises, you may want 
```suggestion
@inline function geom_mapping(D, ::Type{T}) where T
```

But I'm not sure about constant propagation for `D`. We should make sure this is type stable in real uses.

evetion replied:
julia> @time StructTypes.subtypes(GeoJSON.GeoJSONT{2,Float64})
  0.000001 seconds
(Point = GeoJSON.Point{2, Float64}, MultiPoint = GeoJSON.MultiPoint{2, Float64}, LineString = GeoJSON.LineString{2, Float64}, MultiLineString = GeoJSON.MultiLineString{2, Float64}, Polygon = GeoJSON.Polygon{2, Float64}, MultiPolygon = GeoJSON.MultiPolygon{2, Float64}, GeometryCollection = GeoJSON.GeometryCollection{2, Float64}, Feature = GeoJSON.Feature{2, Float64}, FeatureCollection = GeoJSON.FeatureCollection{2, Float64})

And @inferred works. What would you want to improve on this?

rafaqz replied:
Have you checked for a clean profile with ProfileView.jl? `@inferred` lies because it's just checking the output type is known, there can will be internal type stability.

Its not important I'm just going for worlds fastest GeoJSON reader now :smile:

evetion replied:
Squicky clean! There's probably improvements elsewhere, but we can improve them later without breaking anything.

rafaqz commented on src/geojson_types.jl:
```diff
@@ -0,0 +1,350 @@
+    MultiPolygon{D,T}(; bbox=nothing, coordinates=nothing) where {D,T} = MultiPolygon{D,T}(bbox, coordinates)
+end
+MultiPolygon(; bbox=nothing, coordinates::Vector{Vector{Vector{NTuple{2,T}}}}) where {T} = MultiPolygon{2,T}(bbox, coordinates)
+Base.eltype(::Type{MultiPolygon{D,T}}) where {D,T} = Vector{Vector{NTuple{D,T}}}
```
What is `eltype` for ?

And if you are dispatching on `Vector` anyway why not have another method that catches `D` ?

```julia
MultiPolygon(; bbox=nothing, coordinates::Vector) where {T} = 
    MultiPolygon(bbox, coordinates)

MultiPolygon(bbox, coordinates::Vector{Vecor{NTuple{D,T}) where {D,T} = 
    MultiPolygon(bbox, coordinates::Vector{Vector{NTuple{D,T}}})
```

Then 2d and 3d work.

rafaqz replied:
Just updated the example for a typeo, this wont hit unbound args now

evetion replied:
Hahaha, I wish, but that still throws the test error.

rafaqz replied:
Haha wild maybe NTuple includes `D == 0` so there is no T ?? can you Make a union where D is 2, 3, or 4 ??

evetion replied:
Can I propose we either fix this later, or you try to play around with these types? I'm not even sure how to make D a Union for {2,3,4}.

rafaqz replied:
Yeah its annoying with numbers, you need to make a Union over all the types with contained numbers rather than just `{2,3,4}`.

I can PR that - my concern was that it may be a breaking change.

rafaqz replied:
Actually lets just merge this, I need to get the other PRs merged too, like writing tables.

rafaqz commented on test/runtests.jl:
```diff
@@ -237,51 +221,52 @@ include("geojson_samples.jl")
         # "c" and "d" are only present in the properties of a single row
         @test all(t.c .=== ["only-here", missing, missing])
         @test all(t.d .=== [missing, "appears-later", missing])
         @testset "With NamedTuple feature" begin
```
Its not a NamedTuple now this is a bit confusing.

Does `NamedTuple` as properties not work now?

evetion replied:
Ok, now switched to NamedTuple for the fun of it. Can always revert.

rafaqz commented on test/geojson_samples.jl:
```diff
@@ -1,5 +1,4 @@
-
-module T
+module Samples
```
:partying_face: :tada: :100:

rafaqz commented on test/runtests.jl:
```diff
@@ -9,160 +9,150 @@ using Test
             [:type => "é"],
             [:type => "meow"],
             [:title => "Dict 1"],
-            [
```
Why does it need `Any` now?

rafaqz reviewed (changes requested):
Mostly great, some comments on types and minor improvments.

rafaqz commented on test/runtests.jl:
```diff
@@ -9,160 +9,150 @@ using Test
-                :link => "http://example.org/features/1",
-                :summary => "The first feature",
-                :title => "Feature 1",
+            Pair{Symbol,Any}[
```
Why does this need `Any` now?

rafaqz commented on test/runtests.jl:
```diff
@@ -174,61 +164,55 @@ include("geojson_samples.jl")
-        @test geom[1][1][2] == [-117.907767, 33.967747]
-        @test geom[1][1][3] == [-117.912919, 33.96445]
-        @test geom[1][1][4] == [-117.913883, 33.96657]
+        @test geom[1][1][1] == (-117.913883f0, 33.96657f0)
```
So happy that these are `Tuple` now!!

rafaqz commented:
Seems that write here no longer accepts other geometries? It needs to write anythign not just GeoJSON objects

visr commented on src/geojson_types.jl:
```diff
@@ -0,0 +1,340 @@
+"""
+    LazyFeatureCollection{D,T}(bbox::Union{Nothing,Vector{T}}, features::Vector{LazyFeature{D,T}}, crs::Union{Nothing,String})
+
+A FeatureCollection with `D` dimensional geometry in its features, but it's features are
```
```suggestion
A FeatureCollection with `D` dimensional geometry in its features, but its features are
```

visr commented on src/io.jl:
```diff
@@ -0,0 +1,90 @@
+"""
+    GeoJSON.read(json::String; lazyfc=false, ndim=2)
+
+Read GeoJSON from a string or file to a GeoInterface.jl compatible object.
```
I tried a file but it doesn't work yet since https://github.com/quinnj/JSON3.jl/pull/243 is not yet tagged.

visr commented on src/io.jl:
```diff
@@ -0,0 +1,90 @@
+When reading in huge featurecollections (1M+ features), set `lazyfc=true`
+to only parse them into memory when accessed.
+"""
+function read(io; lazyfc=false, ndim=2, numbertype=Float32)
```
Shouldn't we default to numbertype Float64? The numbertype is missing from the docstring.
Also perhaps we should rename `lazyfc` to `lazy`, and just document that it only works for FeatureCollections.

visr replied:
I see you commented on the reason for Float32 [here](https://github.com/JuliaGeo/GeoJSON.jl/pull/63#issuecomment-1400133216). The number of decimals to store in the GeoJSON is discussed here: https://www.rfc-editor.org/rfc/rfc7946#section-11.2. I would lean to using a safe default of Float64, but am fine with Float32 as well.

rafaqz replied:
https://xkcd.com/2170/

visr replied:
Hehe yeah I'm aware that for lat/lon it doesn't make much sense. I'm thinking more about projected coordinates, which can also be used, the spec mentions:

> However, where all involved parties have a prior arrangement, alternative coordinate reference systems can be used without risk of data being misinterpreted.

But even then a typical dutch y coordinate gives 3cm precision (`eps(400_000.0f0) => 0.03125f0`), which for most usecases should be fine. And if it's not it's easy to request more precision with `numbertype`.

rafaqz commented:
I've been using this branch for day to day work, but hitting some bugs.

For example his file was created and loads properly with `main`, but fails on this branch.
https://gist.github.com/rafaqz/3ba5a1d29df15d2e2a61bb577482a470

evetion commented:
On mobile here; What's the error you get with that GeoJSON?

rafaqz commented:
```julia
┌ Warning: Failed to parse GeoJSON as 2D, trying 3D. Set `ndim` to 3 to avoid this warning.
└ @ GeoJSON ~/.julia/dev/GeoJSON/src/io.jl:18
ERROR: ArgumentError: invalid JSON at byte position 127 while parsing type Float32: InvalidChar
# (commented as it breaks highlighting of the rest) s":[[55.26803,-21.018473],[55.271763,-21.014929],[5

Stacktrace:
  [1] invalid(error::JSON3.Error, buf::Vector{UInt8}, pos::Int64, T::Type)
    @ JSON3 ~/.julia/packages/JSON3/CpNms/src/JSON3.jl:30
  [2] read(::StructTypes.NumberType, buf::Vector{UInt8}, pos::Int64, len::Int64, b::UInt8, ::Type{Float32}; pars
equoted::Bool, kw::Base.Pairs{Symbol, Union{}, Tuple{}, NamedTuple{(), Tuple{}}})
    @ JSON3 ~/.julia/packages/JSON3/CpNms/src/structs.jl:206
  [3] read
    @ ~/.julia/packages/JSON3/CpNms/src/structs.jl:187 [inlined]
  [4] TupleClosure
    @ ~/.julia/packages/JSON3/CpNms/src/structs.jl:299 [inlined]
  [5] construct
    @ ~/.julia/packages/StructTypes/AK4aM/src/StructTypes.jl:620 [inlined]
  [6] #read#37
    @ ~/.julia/packages/JSON3/CpNms/src/structs.jl:336 [inlined]
  [7] read
    @ ~/.julia/packages/JSON3/CpNms/src/structs.jl:322 [inlined]
  [8] #read#30
    @ ~/.julia/packages/JSON3/CpNms/src/structs.jl:209 [inlined]
  [9] read
    @ ~/.julia/packages/JSON3/CpNms/src/structs.jl:209 [inlined]
 [10] readarray(buf::Vector{UInt8}, pos::Int64, len::Int64, b::UInt8, ::Type{Vector{Tuple{Float32, Float32, Floa
t32}}}, ::Type{Tuple{Float32, Float32, Float32}}; kw::Base.Pairs{Symbol, Union{}, Tuple{}, NamedTuple{(), Tuple{
}}})
    @ JSON3 ~/.julia/packages/JSON3/CpNms/src/structs.jl:228
 [11] readarray
    @ ~/.julia/packages/JSON3/CpNms/src/structs.jl:213 [inlined]
 [12] #read#31
    @ ~/.julia/packages/JSON3/CpNms/src/structs.jl:210 [inlined]
 [13] read
    @ ~/.julia/packages/JSON3/CpNms/src/structs.jl:210 [inlined]
 [14] #read#30
    @ ~/.julia/packages/JSON3/CpNms/src/structs.jl:209 [inlined]
 [15] read
    @ ~/.julia/packages/JSON3/CpNms/src/structs.jl:209 [inlined]
 [16] macro expansion
    @ ~/.julia/packages/JSON3/CpNms/src/structs.jl:77 [inlined]
 [17] #read#22
    @ ~/.julia/packages/JSON3/CpNms/src/structs.jl:69 [inlined]
 [18] read
    @ ~/.julia/packages/JSON3/CpNms/src/structs.jl:69 [inlined]
 [19] #_#49
    @ ~/.julia/packages/JSON3/CpNms/src/structs.jl:556 [inlined]
 [20] StructClosure
    @ ~/.julia/packages/JSON3/CpNms/src/structs.jl:554 [inlined]
 [21] applyfield
    @ ~/.julia/packages/StructTypes/AK4aM/src/StructTypes.jl:872 [inlined]
 [22] read(::StructTypes.UnorderedStruct, buf::Vector{UInt8}, pos::Int64, len::Int64, b::UInt8, ::Type{GeoJSON.L
ineString{3, Float32}}; kw::Base.Pairs{Symbol, Union{}, Tuple{}, NamedTuple{(), Tuple{}}})
    @ JSON3 ~/.julia/packages/JSON3/CpNms/src/structs.jl:617
 [23] read(::StructTypes.UnorderedStruct, buf::Vector{UInt8}, pos::Int64, len::Int64, b::UInt8, ::Type{GeoJSON.L
ineString{3, Float32}})
    @ JSON3 ~/.julia/packages/JSON3/CpNms/src/structs.jl:566
 [24] #_read#54
    @ ~/.julia/packages/JSON3/CpNms/src/structs.jl:804 [inlined]
 [25] _read
    @ ~/.julia/packages/JSON3/CpNms/src/structs.jl:748 [inlined]
 [26] #read#53
    @ ~/.julia/packages/JSON3/CpNms/src/structs.jl:745 [inlined]
 [27] read
    @ ~/.julia/packages/JSON3/CpNms/src/structs.jl:739 [inlined]
 [28] macro expansion
    @ ~/.julia/packages/JSON3/CpNms/src/structs.jl:77 [inlined]
 [29] read(::StructTypes.UnorderedStruct, buf::Vector{UInt8}, pos::Int64, len::Int64, b::UInt8, T::Type{Union{No
thing, GeoJSON.AbstractGeometry{3, Float32}}}; kw::Base.Pairs{Symbol, Union{}, Tuple{}, NamedTuple{(), Tuple{}}}
)
    @ JSON3 ~/.julia/packages/JSON3/CpNms/src/structs.jl:69
 [30] read
    @ ~/.julia/packages/JSON3/CpNms/src/structs.jl:69 [inlined]
 [31] #_#49
    @ ~/.julia/packages/JSON3/CpNms/src/structs.jl:556 [inlined]
 [32] StructClosure
    @ ~/.julia/packages/JSON3/CpNms/src/structs.jl:554 [inlined]
 [33] applyfield
    @ ~/.julia/packages/StructTypes/AK4aM/src/StructTypes.jl:872 [inlined]
 [34] read(::StructTypes.UnorderedStruct, buf::Vector{UInt8}, pos::Int64, len::Int64, b::UInt8, ::Type{GeoJSON.F
eature{3, Float32}}; kw::Base.Pairs{Symbol, Union{}, Tuple{}, NamedTuple{(), Tuple{}}})
    @ JSON3 ~/.julia/packages/JSON3/CpNms/src/structs.jl:617
 [35] read
    @ ~/.julia/packages/JSON3/CpNms/src/structs.jl:566 [inlined]
 [36] readarray(buf::Vector{UInt8}, pos::Int64, len::Int64, b::UInt8, ::Type{Vector{GeoJSON.Feature{3, Float32}}
}, ::Type{GeoJSON.Feature{3, Float32}}; kw::Base.Pairs{Symbol, Union{}, Tuple{}, NamedTuple{(), Tuple{}}})
    @ JSON3 ~/.julia/packages/JSON3/CpNms/src/structs.jl:228
 [37] readarray
    @ ~/.julia/packages/JSON3/CpNms/src/structs.jl:213 [inlined]
 [38] #read#31
    @ ~/.julia/packages/JSON3/CpNms/src/structs.jl:210 [inlined]
 [39] read
    @ ~/.julia/packages/JSON3/CpNms/src/structs.jl:210 [inlined]
 [40] #read#30
    @ ~/.julia/packages/JSON3/CpNms/src/structs.jl:209 [inlined]
 [41] read
    @ ~/.julia/packages/JSON3/CpNms/src/structs.jl:209 [inlined]
 [42] #_#49
    @ ~/.julia/packages/JSON3/CpNms/src/structs.jl:556 [inlined]
 [43] StructClosure
    @ ~/.julia/packages/JSON3/CpNms/src/structs.jl:554 [inlined]
 [44] applyfield
    @ ~/.julia/packages/StructTypes/AK4aM/src/StructTypes.jl:872 [inlined]
 [45] read(::StructTypes.UnorderedStruct, buf::Vector{UInt8}, pos::Int64, len::Int64, b::UInt8, ::Type{GeoJSON.F
eatureCollection{3, Float32}}; kw::Base.Pairs{Symbol, Union{}, Tuple{}, NamedTuple{(), Tuple{}}})
    @ JSON3 ~/.julia/packages/JSON3/CpNms/src/structs.jl:617
 [46] read(::StructTypes.UnorderedStruct, buf::Vector{UInt8}, pos::Int64, len::Int64, b::UInt8, ::Type{GeoJSON.F
eatureCollection{3, Float32}})
    @ JSON3 ~/.julia/packages/JSON3/CpNms/src/structs.jl:566
 [47] #_read#54
    @ ~/.julia/packages/JSON3/CpNms/src/structs.jl:804 [inlined]
 [48] _read
    @ ~/.julia/packages/JSON3/CpNms/src/structs.jl:748 [inlined]
 [49] #read#53
    @ ~/.julia/packages/JSON3/CpNms/src/structs.jl:745 [inlined]
 [50] read
    @ ~/.julia/packages/JSON3/CpNms/src/structs.jl:739 [inlined]
 [51] #read#44
    @ ~/.julia/packages/JSON3/CpNms/src/structs.jl:433 [inlined]
 [52] read
    @ ~/.julia/packages/JSON3/CpNms/src/structs.jl:431 [inlined]
 [53] read(str::JSON3.VectorString{Vector{UInt8}}, ::Type{GeoJSON.GeoJSONWrapper{3, Float32, X} where X<:GeoJSON
.GeoJSONT{3, Float32}}; jsonlines::Bool, kw::Base.Pairs{Symbol, Union{}, Tuple{}, NamedTuple{(), Tuple{}}})
    @ JSON3 ~/.julia/packages/JSON3/CpNms/src/structs.jl:40
 [54] read
    @ ~/.julia/packages/JSON3/CpNms/src/structs.jl:32 [inlined]
 [55] #read#15
    @ ~/.julia/packages/JSON3/CpNms/src/structs.jl:14 [inlined]
 [56] read(bytes::Vector{UInt8}, ::Type{GeoJSON.GeoJSONWrapper{3, Float32, X} where X<:GeoJSON.GeoJSONT{3, Float
32}})
    @ JSON3 ~/.julia/packages/JSON3/CpNms/src/structs.jl:14
 [57] read(io::Vector{UInt8}; lazyfc::Bool, ndim::Int64, numbertype::Type)
    @ GeoJSON ~/.julia/dev/GeoJSON/src/io.jl:19
 [58] read(io::Vector{UInt8})
    @ GeoJSON ~/.julia/dev/GeoJSON/src/io.jl:10

caused by: ArgumentError: invalid JSON at byte position 354 while parsing type JSON3.Object: ExpectedOpeningObje
ctChar
# (commented as it breaks highlighting of the rest) 0.987873]]},"properties":null},{"type":"Feature","g

Stacktrace:
  [1] invalid(error::JSON3.Error, buf::Vector{UInt8}, pos::Int64, T::Type)
    @ JSON3 ~/.julia/packages/JSON3/CpNms/src/JSON3.jl:30
  [2] #read#43
    @ ~/.julia/packages/JSON3/CpNms/src/structs.jl:428 [inlined]
  [3] read
    @ ~/.julia/packages/JSON3/CpNms/src/structs.jl:352 [inlined]
  [4] #read#39
    @ ~/.julia/packages/JSON3/CpNms/src/structs.jl:347 [inlined]
  [5] read
    @ ~/.julia/packages/JSON3/CpNms/src/structs.jl:347 [inlined]
  [6] #_#49
    @ ~/.julia/packages/JSON3/CpNms/src/structs.jl:556 [inlined]
  [7] StructClosure
    @ ~/.julia/packages/JSON3/CpNms/src/structs.jl:554 [inlined]
  [8] applyfield
    @ ~/.julia/packages/StructTypes/AK4aM/src/StructTypes.jl:872 [inlined]
  [9] read(::StructTypes.UnorderedStruct, buf::Vector{UInt8}, pos::Int64, len::Int64, b::UInt8, ::Type{GeoJSON.F
eature{2, Float32}}; kw::Base.Pairs{Symbol, Union{}, Tuple{}, NamedTuple{(), Tuple{}}})
    @ JSON3 ~/.julia/packages/JSON3/CpNms/src/structs.jl:617
 [10] read
    @ ~/.julia/packages/JSON3/CpNms/src/structs.jl:566 [inlined]
 [11] readarray(buf::Vector{UInt8}, pos::Int64, len::Int64, b::UInt8, ::Type{Vector{GeoJSON.Feature{2, Float32}}
}, ::Type{GeoJSON.Feature{2, Float32}}; kw::Base.Pairs{Symbol, Union{}, Tuple{}, NamedTuple{(), Tuple{}}})
    @ JSON3 ~/.julia/packages/JSON3/CpNms/src/structs.jl:228
 [12] readarray
    @ ~/.julia/packages/JSON3/CpNms/src/structs.jl:213 [inlined]
 [13] #read#31
    @ ~/.julia/packages/JSON3/CpNms/src/structs.jl:210 [inlined]
 [14] read
    @ ~/.julia/packages/JSON3/CpNms/src/structs.jl:210 [inlined]
 [15] #read#30
    @ ~/.julia/packages/JSON3/CpNms/src/structs.jl:209 [inlined]
 [16] read
    @ ~/.julia/packages/JSON3/CpNms/src/structs.jl:209 [inlined]
 [17] #_#49
    @ ~/.julia/packages/JSON3/CpNms/src/structs.jl:556 [inlined]
 [18] StructClosure
    @ ~/.julia/packages/JSON3/CpNms/src/structs.jl:554 [inlined]
 [19] applyfield
    @ ~/.julia/packages/StructTypes/AK4aM/src/StructTypes.jl:872 [inlined]
 [20] read(::StructTypes.UnorderedStruct, buf::Vector{UInt8}, pos::Int64, len::Int64, b::UInt8, ::Type{GeoJSON.F
eatureCollection{2, Float32}}; kw::Base.Pairs{Symbol, Union{}, Tuple{}, NamedTuple{(), Tuple{}}})
    @ JSON3 ~/.julia/packages/JSON3/CpNms/src/structs.jl:617
 [21] read(::StructTypes.UnorderedStruct, buf::Vector{UInt8}, pos::Int64, len::Int64, b::UInt8, ::Type{GeoJSON.F
eatureCollection{2, Float32}})
    @ JSON3 ~/.julia/packages/JSON3/CpNms/src/structs.jl:566
 [22] #_read#54
    @ ~/.julia/packages/JSON3/CpNms/src/structs.jl:804 [inlined]
 [23] _read
    @ ~/.julia/packages/JSON3/CpNms/src/structs.jl:748 [inlined]
 [24] #read#53
    @ ~/.julia/packages/JSON3/CpNms/src/structs.jl:745 [inlined]
 [25] read
    @ ~/.julia/packages/JSON3/CpNms/src/structs.jl:739 [inlined]
 [26] #read#44
    @ ~/.julia/packages/JSON3/CpNms/src/structs.jl:433 [inlined]
 [27] read
    @ ~/.julia/packages/JSON3/CpNms/src/structs.jl:431 [inlined]
 [28] read(str::JSON3.VectorString{Vector{UInt8}}, ::Type{GeoJSON.GeoJSONWrapper{2, Float32, X} where X<:GeoJSON
.GeoJSONT{2, Float32}}; jsonlines::Bool, kw::Base.Pairs{Symbol, Union{}, Tuple{}, NamedTuple{(), Tuple{}}})
    @ JSON3 ~/.julia/packages/JSON3/CpNms/src/structs.jl:40
 [29] read
    @ ~/.julia/packages/JSON3/CpNms/src/structs.jl:32 [inlined]
 [30] #read#15
    @ ~/.julia/packages/JSON3/CpNms/src/structs.jl:14 [inlined]
 [31] read(bytes::Vector{UInt8}, ::Type{GeoJSON.GeoJSONWrapper{2, Float32, X} where X<:GeoJSON.GeoJSONT{2, Float
32}})
    @ JSON3 ~/.julia/packages/JSON3/CpNms/src/structs.jl:14
 [32] read(io::Vector{UInt8}; lazyfc::Bool, ndim::Int64, numbertype::Type)
    @ GeoJSON ~/.julia/dev/GeoJSON/src/io.jl:15
 [33] read(io::Vector{UInt8})
    @ GeoJSON ~/.julia/dev/GeoJSON/src/io.jl:10
 [34] top-level scope
    @ REPL[6]:1
```

rafaqz commented:
Maybe its properties being `null` ?

evetion commented:
> Maybe its properties being `null` ?

Yep, that's it. Should be `{Nothing, NamedTuple}`, now it's just `NamedTuple`.

rafaqz commented:
Sweet easy fix

evetion commented:
> I've been using this branch for day to day work, but hitting some bugs.

Can you describe the other bugs? Otherwise, especially if you can use this for your day to day work, let's merge this to main, revive #59 and make a breaking release.

rafaqz commented:
One waa it's really slow using mixed geometry type features from overpass turbo. I'll have to pit an MWE together.

Something else didn't work but it might have been `null` again.

rafaqz commented:
Ok MWE of slow loads. I chose a workload that would take a second or so so we can profile it. My use case needs the whole island, which takes a lot longer.

The file is here: https://drive.google.com/file/d/1B-3ziH0oWivLye6V0wKk40Nxr3PTsAb0/view?usp=sharing

```julia
using GeoJSON
using ProfileView
@time GeoJSON.read(read("slow.geojson"))
```

Gives:
```julia
julia> @time GeoJSON.read(read("slow.geojson"))
  1.430720 seconds (1.13 M allocations: 54.966 MiB)
FeatureCollection with 5789 Features
```

And the profile:
```julia
@profview GeoJSON.read(read("slow.geojson"))
```

Mostly type instability in StructTypes.jl
![2023-03-06-104518_1049x844](https://user-images.githubusercontent.com/2534009/223074440-f03ed6b6-ef8f-4b39-9080-c40ef08749f7.png)

It may be better if the properties are a `Dict`? Still wont be fast, but better. The more I look at it the more I hate GeoJSON as a format.

evetion commented:
Yeah, the format isn't that great. Anyway, slow.geojson isn't that slow for me. Oddly, when profiling, I seem to arrive at the `invalid` method, while no error is thrown:
<img width="353" alt="image" src="https://user-images.githubusercontent.com/8655030/223431097-6896b01c-7c5e-4109-8db6-65e007c50018.png">

Not sure if that's an error in my installation. To be fair, changing the code back to Dict doesn't seem to do much in terms of time, and only a little bit in terms of allocations.

rafaqz commented:
How slow is "isn't that slow" ?

Are you on julia 1.9? That's a very different profile it would be good to know whats happening here.

rafaqz commented:
I think the other half of my profile was that "invalid" method. So you just have that part, without the StructTypes instabilities. Interesting.

evetion commented:
> How slow is "isn't that slow" ?

Properties made into Dict again

1.9
```julia
julia> @time GeoJSON.read(data)
  1.686686 seconds (3.22 M allocations: 157.501 MiB, 4.47% gc time, 64.26% compilation time)
FeatureCollection with 5789 Features

julia> @time GeoJSON.read(data)
  0.600645 seconds (983.32 k allocations: 39.599 MiB, 1.28% gc time)
FeatureCollection with 5789 Features
```

1.8
```
julia> @time GeoJSON.read(data)
  2.148807 seconds (3.77 M allocations: 174.611 MiB, 3.27% gc time, 69.70% compilation time)
FeatureCollection with 5789 Features

julia> @time GeoJSON.read(data)
  0.646021 seconds (936.02 k allocations: 38.156 MiB, 5.79% gc time)
FeatureCollection with 5789 Features
```

---

this PR (still with NamedTuple)

1.9
```julia
julia> @time GeoJSON.read(data)
  7.736202 seconds (10.49 M allocations: 623.268 MiB, 1.06% gc time, 87.94% compilation time)
FeatureCollection with 5789 Features

julia> @time GeoJSON.read(data)
  0.806888 seconds (1.15 M allocations: 46.902 MiB, 1.44% gc time)
FeatureCollection with 5789 Features
```

1.8
```julia
julia> @time GeoJSON.read(data)
  8.658965 seconds (14.97 M allocations: 787.488 MiB, 1.99% gc time, 88.66% compilation time)
FeatureCollection with 5789 Features

julia> @time GeoJSON.read(data)
  0.821887 seconds (1.10 M allocations: 44.540 MiB, 0.69% gc time)
FeatureCollection with 5789 Features
```


---

So it seems that while the NamedTuple allocates slightly more, it is mostly much harder on type inference, the first call is much much slower.

evetion commented:
> I think the other half of my profile was that "invalid" method. So you just have that part, without the StructTypes instabilities. Interesting.

Also, I don't understand why it would hit `invalid`, it's basicly everything in my profiles (whether Dict or NamedTuple) and this is the code:

```julia
@noinline invalid(error, buf, pos, T) = throw(ArgumentError("""
invalid JSON at byte position $pos while parsing type $T: $error
$(String(buf[max(1, pos-25):min(end, pos+25)]))
"""))
```

rafaqz commented:
Nice, so `Dict` it is!

But that `invalid` is super weird, how does it take all time but not throw visibly??

Is it getting caught by a try/catch block somewhere and ignored?

rafaqz commented:
I guess its this try - catch:

https://github.com/quinnj/JSON3.jl/blob/main/src/structs.jl#L82-L86

We could remove the block and see what the error is.

evetion commented:
> I guess its this try - catch:
> 
> [quinnj/JSON3.jl@`main`/src/structs.jl#L82-L86](https://github.com/quinnj/JSON3.jl/blob/main/src/structs.jl?rgh-link-date=2023-03-07T20%3A02%3A45Z#L82-L86)
> 
> We could remove the try catch and just run the first `read` and see what the error is.

Yes and it's caused by the id of the Feature, having a unholy Union. 

Changing from `id::Union{Nothing,Int,String}` to `id::Union{Nothing,String}` you get:

```julia
julia> @time GeoJSON.read(data)
  1.174992 seconds (3.05 M allocations: 148.982 MiB, 2.30% gc time, 91.93% compilation time)
FeatureCollection with 5789 Features

julia> @time GeoJSON.read(data)
  0.086503 seconds (890.70 k allocations: 35.536 MiB)
FeatureCollection with 5789 Features
```

Spec however says:
>    o  If a Feature has a commonly used identifier, that identifier
      SHOULD be included as a member of the Feature object with the name
      "id", and the value of this member is either a JSON string or
      number.

rafaqz commented:
I would burn the spec for that speedup, especially if it just means ids are always strings.

But can we hack around it some how by defining a custom `read`?

evetion commented:
> I would burn the spec for that speedup, especially if it just means ids are always strings.
> 
> But can we hack around it some how by defining a custom `read`?
> 
> Where we check `isnumeric`?

Or setting it to Any? Ugly, but just as fast.

rafaqz commented:
Oh yeah let's do that. But does it actually parse the numbers? Or just make them all strings anyway.

evetion commented:
> Oh yeah let's do that. But does it actually parse the numbers? Or just make them all strings anyway.

```julia
julia> d = """{
           "type": "Feature",
           "id": 1,
           "bbox": [-180.0, -90.0, 180.0, 90.0],
           "geometry": {"type": "MultiLineString", "coordinates": [[[3.75, 9.25], [-130.95, 1.52]], [[23.15, -34.25], [-1.35, -4.65], [3.45, 77.95]]]},
           "properties": {"title": "Dict 1"}
       }""";

julia> f = GeoJSON.read(d)
Feature with 2D MultiLineString geometry and 2 properties: (:geometry, :title)

julia> f.id
1

julia> typeof(f.id)
Int64
```

String also works. I did note that accessing the properties is a bit type unstable, due to us overloading getproperty, but it still seems pretty fast to me.

Only some culling of the precompile is needed. While the large examples (yours, the original Makie one), now take ~1s to parse for the first time and ~0.05s for subsequent calls, precompiling now takes ~15s.

rafaqz commented:
I'm fine with shifting that time to precompile for that performance gain. SnoopCompile `@snoopi_deep` might show some obvious fixable problems too.

But we should just get this merged, I've already switched to it for practical work, the precompile is less than it was taking to load single json files.

evetion commented:
Feel free to merge

rafaqz reviewed (approved):
Lgtm. 

Im wondering how many of those precompile statements actually work in practice, just running tje functions tends to be better if there is any type instability. But we can fix that later.
